prover/src/gpu.rs

```diff
@@ -0,0 +1,284 @@
+use winter_prover::{crypto::MerkleTree, ColMatrix, Prover, RowMatrix, StarkDomain};
+
+const RPO_RATE: usize = Rpo256::RATE_RANGE.end - Rpo256::RATE_RANGE.start;
+
+/// Wraps an [ExecutionProver] and provides GPU acceleration for building Rpo256 trace commitments.
+pub(crate) struct GpuRpoExecutionProver<R>(pub ExecutionProver<Rpo256, R>)
+where
+    R: RandomCoin<BaseField = Felt, Hasher = Rpo256>;
+
+impl<R> Prover for GpuRpoExecutionProver<R>
+where
+    R: RandomCoin<BaseField = Felt, Hasher = Rpo256>,
+{
+    type Air = <ExecutionProver<Rpo256, R> as Prover>::Air;
+    type BaseField = Felt;
+    type Trace = <ExecutionProver<Rpo256, R> as Prover>::Trace;
+    type HashFn = Rpo256;
+    type RandomCoin = R;
+
+    fn options(&self) -> &WinterProofOptions {
+        self.0.options()
+    }
+
+    fn get_pub_inputs(&self, trace: &ExecutionTrace) -> PublicInputs {
+        self.0.get_pub_inputs(trace)
+    }
+
+    /// Computes a low-degree extension (LDE) of the provided execution trace over the specified
+    /// domain and builds a commitment to the extended trace.
+    ///
+    /// The extension is performed by interpolating each column of the execution trace into a
+    /// polynomial of degree = trace_length - 1, and then evaluating the polynomial over the LDE
+    /// domain.
+    ///
+    /// Trace commitment is computed by hashing each row of the extended execution trace, and then
+    /// building a Merkle tree from the resulting hashes.
+    ///
+    /// Interpolations and evaluations are computed on the CPU while hashes are simultaneously
+    /// computed on the GPU:
+    ///
+    /// ```text
+    ///        ──────────────────────────────────────────────────────
+    ///               ┌───┐   ┌────┐   ┌───┐   ┌────┐   ┌───┐
+    ///  CPU:   ... ──┤fft├─┬─┤ifft├───┤fft├─┬─┤ifft├───┤fft├─┬─ ...
+    ///               └───┘ │ └────┘   └───┘ │ └────┘   └───┘ │
+    ///        ╴╴╴╴╴╴╴╴╴╴╴╴╴┼╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴┼╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴╴┼╴╴╴╴╴╴
+    ///                     │ ┌──────────┐   │ ┌──────────┐   │
+    ///  GPU:               └─┤   hash   │   └─┤   hash   │   └─ ...
+    ///                       └──────────┘     └──────────┘
+    ///        ────┼────────┼────────┼────────┼────────┼────────┼────
+    ///           t=n     t=n+1    t=n+2     t=n+3   t=n+4    t=n+5
+    /// ```
+    fn build_trace_commitment<E>(
+        &self,
+        trace: &ColMatrix<E>,
+        domain: &StarkDomain<Felt>,
+    ) -> (RowMatrix<E>, MerkleTree<Self::HashFn>, ColMatrix<E>)
+    where
+        E: air::FieldElement<BaseField = Felt>,
+    {
+        // interpolate the execution trace
+        let now = Instant::now();
+        let poly_size = trace.num_rows();
+        let inv_twiddles = fft::get_inv_twiddles::<E::BaseField>(trace.num_rows());
+        let trace_polys = trace.columns().map(|col| {
+            let mut poly = col.to_vec();
+            fft::interpolate_poly(&mut poly, &inv_twiddles);
+            poly
+        });
```

Comment:
I think we should be able to do this as:
```Rust
let trace_polys = trace.interpolate_columns();
```
This would also give us a `ColMatrix` back - which I think will be convenient later.

Reply:
Happy to interpolate the columns all upfront. It would simplify the code but also have a minor impact on performance. For instance `cargo run --release --features executable,metal,concurrent -- example -r fib -n 131072` takes an extra 500ms to generate the proof on my machine with this approach.

Reply:
Ah - it all makes sense now! Let's keep it as you have now - much better than the alternative.